cpu/atmega_common: pseudomodule-based pin change interrupt implementation

roberthartung opened the pull request:
### Contribution description

This is pseudomodule-based version of pin change interrupts for the atmega common platform. Challenge was, that each bank of pin change interrupts (PCINT0, ..., PCINT3) mixes ports. Therefore some magic has to happen.

The memory consumption is: (1 byte (state) + 8 byte (mapping) + 8*6 (config)) per used bank. This means if a single PCINT is used on a pin we need 57 bytes of additional memory.

Additionally, each cpu needs an `atmega_pcint.h` for the specific PCINT mapping.

### Issues/PRs references

See also #7610, #8993, #9159. This is an follow up PR for #11114 

### Testing
- [x] ATmega1281 (board(s): `waspmote-pro`) @maribu double checked config against the data sheet (no one has the board)
- [x] ATmega1284p (board(s): `mega-xplained`) @TorbenPetersen tested with the INGA board
- [x] ATmega2560 (boards(s): `arduino-mega2560`) PCINT9 and PCINT10 are not working yet
- [x] ATmega256RFR2 (boards(s): `jiminy-mega256rfr2`) @maribu double checked config against the data sheet
- [x] ATmega328P (boards(s): `arduino-uno`) @maribu tested, works as expected

#### Testing for non-Arduino boards
1. Flash `tests/periph_gpio` on your avr board
2. Enable a pcint bank with `USEMODULE += atmega_pcint0`, or `USEMODULE += atmega_pcint`.
3. Using the shell commands `init_int` and `enable_int` test if interrupts are correctly fired for a couple of pins.
    - Please test both pins with their own interrupt vector (to rule out regressions) and pins that rely on the "Pin Change Interrupt" vector shared by multiple GPIOs
    - Please test with at least one pin per interrupt bank

#### Testing for Arduino Uno and Arduino Mega2560
Flash and run the test added in this PR.

------

Update 1: Added TODO for testing
Update 2: Elaborated on testing
Update 3: Added instructions for testing on Arduino boards, added test results

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -59,6 +63,88 @@
+/**
+ * @brief
+ */
+static gpio_t pcint_mapping[][8] = {
```
This can be `const`, can't it?

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -195,8 +281,70 @@ int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
+    /* not a valid interrupt pin. Set as pcint instead if pcints are enabled */
     if (int_num < 0) {
+	    /* If pin change interrupts are enabled, enable mask and interrupt */
+        #ifdef PCINT_NUM_BANKS
```
We do not have an official RIOT coding style on indent of preprocessor, so treat this as a comment and feel free to just ignore it:

I personally do not favor to keep C language indent for CPP, as those are two different language and cannot reasonably be mixed. E.g. 

```C
switch(var) {
    case FOO:
        /* some code */
        break;
     #idef HAS_BAR
     case BAR:
         /* some code */
         break;
         #endif /* <-- two or one level of indent?
}
```

Therefore I personally favor to have no indent before the preprocessor macros. This also makes them more visible, as those look a bit out of place. This additional "spotlight" on c preprocessor code make overlooking them harder and - in my personal opinion - reading the code easier.

(But I again: Feel free to ignore this, there is no policy on that a.f.a.i.k.)

roberthartung replied:
no, you're right! :)

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -195,8 +281,70 @@ int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
+	    /* If pin change interrupts are enabled, enable mask and interrupt */
+        #ifdef PCINT_NUM_BANKS
+        int8_t offset = -1;
+        //uint8_t port_num = _port_num(pin);
```
Forgot to remove comment?

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -195,8 +281,70 @@ int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
+        //uint8_t port_num = _port_num(pin);
+        uint8_t pin_num = _pin_num(pin);
+        uint8_t b;
+        for(b=0; b<PCINT_NUM_BANKS; b++) {
```
Missing space between `for` and `(`. Same in multiple other lines. (I'd suggest to run `uncrustify`)

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -195,8 +281,70 @@ int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
+        for(b=0; b<PCINT_NUM_BANKS; b++) {
+            for(uint8_t p=0; p<8; p++) {
+                if(pin == pcint_mapping[b][p]) {
+                    offset = b*8+p;
```
spaces missing

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -195,8 +281,70 @@ int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
+            for(uint8_t p=0; p<8; p++) {
+                if(pin == pcint_mapping[b][p]) {
+                    offset = b*8+p;
+                    //pcint_lookup[ _port_num(pin) * 8 + pin_num ] = i;
```
Forgot to remove comment?

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -195,8 +281,70 @@ int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
+        int8_t offset = -1;
+        //uint8_t port_num = _port_num(pin);
+        uint8_t pin_num = _pin_num(pin);
+        uint8_t b;
```
I think a longer name than `b` could increase readability of the code

roberthartung replied:
true!

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -59,6 +63,88 @@
+/**
+ * @brief check which pcints should be enabled!
+ */
+#if defined(MODULE_ATMEGA_PCINT) || defined(MODULE_ATMEGA_PCINT0)
```
You can have module `atmega_pcint` depend on `atmega_pcint0` to `atmega_pcint3` in order to simplify your life here

roberthartung replied:
I am using the ATMEGA_PCINT module to enable *all* pcints at once.

maribu replied:
Using a makefile you can add all the individual bank modules when `atmega_pcint` is used. Then you do not need to handle that in C code

maribu replied:
See comment for `Makefile.include` to see what I meant

roberthartung replied:
But there might be different numbers of banks in CPUs, have to check if this leads to any problems if PCINT3 is enabled but only 0-2 exist.

maribu replied:
True. But I think it is still easier to be handled in the makefile, e.g. using

```
ifneq (,$(filter atmega_pcint,$(USEMODULE))
  USEMODULE += atmega_pcint0 atmega_pcint1 atmega_pcint2
  ifeq (atmega1284p,$(CPU))
    USEMODULE += atmega_pcint3
  endif
endif
```

maribu replied:
Maybe it might also make sense to use the `FEATURES_PROVIDED` and `FEATURES_REQUIRED` variables to guard against misuse. (E.g. a user might manually add `USEMODULE += atmega_pcint3` on an ATmega328p.) If the CPUs add `atmega_pcint<NUM>` for each supported vector to `FEATURES_PROVIDED` and the modules `atmega_pcint<NUM>` add this to `FEATURES_REQUIRED`, the build system will print a human readable error message when someone tries to use pin change interrupts for a non-existent bank

roberthartung replied:
I personally don't like the solution from above. This adds a lot of code specific code to the CPU. In the end, this isn't a problem, because I solved it directly in the code when checking if the bank is enabled: I just check if the mapping exists, if so, the CPU has the bank. Problem solved :)

roberthartung replied:
Working on the FEATURES_PROVIDED/FEATURES_REQUIRED solution right now.

maribu replied:
OK, I'm convinced. But maybe it would improve debugging when instead of

```C 
#if defined(MODULE_ATMEGA_PCINT1) && defined(ATMEGA_PCINT_MAP_PCINT1)
...
#endif
```

this is used:


```C 
#ifdef MODULE_ATMEGA_PCINT1
#ifndef ATMEGA_PCINT_MAP_PCINT1
#error "Either mapping for pin change interrupt bank 1 is missing or not supported by the MCU"
#endif /* ATMEGA_PCINT_MAP_PCINT1 */
...

#endif /* MODULE_ATMEGA_PCINT1 */
...
#endif
```

(This would indeed require cpu-depend handling of `atmega_pcint`, but uses will now what went wrong when the manually added `atmega_pcint<NUM>`

roberthartung replied:
Okay ... had to fix the dependencies, but it seems to be working now!

roberthartung replied:
True, will add some error messages for that!

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -195,8 +281,70 @@ int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
+        gpio_init(pin, mode);
+        /* configure pcint */
+        cli();
+        switch (b) {
```
Shouldn't the cases below be placed in `#ifdef MODULE_PCINT<NUM>` ... `#endif`?

roberthartung replied:
absolutely. And was also buggy until now!

maribu commented on cpu/atmega_common/Makefile.include:
```diff
@@ -6,4 +6,7 @@ export INCLUDES += -I$(RIOTCPU)/atmega_common/include \
 USEMODULE += avr_libc_extra
 
+PSEUDOMODULES += atmega_pcint
+PSEUDOMODULES += atmega_pcint%
```
Just add

```
ifneq (,$(filter atmega_pcint,$(USEMODULE))
    USEMODULE += atmega_pcint0 atmega_pcint1 atmega_pcint2 atmega_pcint3
endif
```

And you no longer need to handle `MODULE_ATMEGA_PCINT` in c code.

roberthartung replied:
See comment above. Will try.

maribu replied:
This should be moved to `makesfiles/pseudomodules.inc.mk`

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -254,6 +405,66 @@ static inline void irq_handler(uint8_t int_num)
+{
+    pcint_handler(PCINT1_IDX, &PCMSK1);
+}
+#endif  /* PCINT1_vect */
```
comment needs to be updated (more occurrences below). Please also add the comment for `PCINT0_IDX` for consistency.

roberthartung replied:
Ack.

maribu commented on cpu/atmega1284p/include/atmega_pcint.h:
```diff
@@ -0,0 +1,7 @@
+#ifndef ATMEGA_PCINT_H
+#define ATMEGA_PCINT_H
+#define ATMEGA_PCINT_MAP_PCINT0 { GPIO_PIN(0, 0), GPIO_PIN(0, 1), GPIO_PIN(0, 2), GPIO_PIN(0, 3), GPIO_PIN(0, 4), GPIO_PIN(0, 5), GPIO_PIN(0, 6), GPIO_PIN(0, 7) }
```
Please use `GPIO_PIN(PORT_A, 0)` instead of `GPIO_PIN(0, 0)`. This will make reading it easier, as the code better matches the names used in the datasheet. (E.g. the pins are name PB0, PA7, ...)

roberthartung replied:
absolutely true :)

roberthartung commented on cpu/atmega1284p/include/atmega_pcint.h:
```diff
@@ -0,0 +1,7 @@
+#define ATMEGA_PCINT_MAP_PCINT1 { GPIO_PIN(1, 0), GPIO_PIN(1, 1), GPIO_PIN(1, 2), GPIO_PIN(1, 3), GPIO_PIN(1, 4), GPIO_PIN(1, 5), GPIO_PIN(1, 6), GPIO_PIN(1, 7) }
+#define ATMEGA_PCINT_MAP_PCINT2 { GPIO_PIN(2, 0), GPIO_PIN(2, 1), GPIO_PIN(2, 2), GPIO_PIN(2, 3), GPIO_PIN(2, 4), GPIO_PIN(2, 5), GPIO_PIN(2, 6), GPIO_PIN(2, 7) }
+#define ATMEGA_PCINT_MAP_PCINT3 { GPIO_PIN(3, 0), GPIO_PIN(3, 1), GPIO_PIN(3, 2), GPIO_PIN(3, 3), GPIO_PIN(3, 4), GPIO_PIN(3, 5), GPIO_PIN(3, 6), GPIO_PIN(3, 7) }
+#endif
```
We need the special case anyway, and the current implementation is even better (memory-wise).
Should we move the definitions to the cpu.h anyway?

maribu replied:
> We need the special case anyway

Yes, and also for non-ATmega2560 while there is a 1:1 mapping for GPIO port and PCINT bank, these mapping are different. I'd personally often prefer less complexity over RAM/ROM improvement.

> the current implementation is even better (memory-wise).

This is a surprise to me. I'd assume in case of the ATmega328p and the ATmega1284p only the port would need to be saved per bank, and the pin can be obtained via `GPIO_PIN(port_of_bank(bank), pin_number)`. But as said, I'd personally would keep it. But feel free to squeeze more ROM bytes out of it, if you see this differently.

> Should we move the definitions to the cpu.h anyway?

I do like the current place. But I do not have a strong opinion about that anyway.

roberthartung commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -254,6 +405,66 @@ static inline void irq_handler(uint8_t int_num)
+            uint8_t pin_mask = (1 << pin_num);
+            gpio_t pin = pcint_mapping[ bank ][ pin_num ];
+            uint8_t port_value = (_SFR_MEM8(_pin_addr( GPIO_PIN( _port_num(pin), pin_num ))));
+            uint8_t pin_value = (port_value & pin_mask) == 0 ? 0 : 1;
```
sounds like a good idea!

maribu commented on cpu/atmega_common/Makefile.dep:
```diff
@@ -0,0 +1,19 @@
+ifneq (,$(filter atmega_pcint,$(USEMODULE)))
+    USEMODULE += atmega_pcint0 atmega_pcint1 atmega_pcint2 atmega_pcint3
```
This will trigger the `#error` in the c file. (I know it is a pitty that it now needs to be handled twice, but I believe the error message can save a lot of time for end users and developers of this feature.)

roberthartung replied:
Don't quite get the comment / problem here? Can you specify your comment?

roberthartung replied:
Oooh. You mean using only atmega_pcint also triggers the atmega_pcint3 error...

maribu commented on cpu/atmega1284p/Makefile.dep:
```diff
@@ -0,0 +1,5 @@
+ifneq (,$(filter atmega_pcint,$(USEMODULE)))
+    USEMODULE += atmega_pcint0 atmega_pcint1 atmega_pcint2
```
I'd put that just into the `Makefile.include` of each individual AVR chip and keep `boards/common/arduino-atmega/Makefile.dep` unchanged and drop this file. Also, the ATmega1284P does have PCINT3 in contrast to the ATmege328P and ATmega2560.

maribu replied:
I also guess that all AVR MCUs have at least PCINT0 to PCINT2, so maybe move the code above to  `cpu/atmega_common/Makefile.include` and just add

```
ifneq (,$(filter atmega_pcint,$(USEMODULE)))
    USEMODULE += atmega_pcint3
endif
```

to `cpu/atmega1284p/Makefile.include` to have only the special case handled at CPU level

maribu replied:
OK, my guess is right at least for the boards supported by RIOT.

ATmega1281, ATmega328P, ATmega2560, ATmega256RFR2: PCINT0 to PCINT2
ATmega1284P: PCINT0 to PCINT3

So I'd just add PCINT0 to PCINT2 for all AVRs and add special case code to the ATmega1284p

roberthartung replied:
Sounds good, will update. I will also check with .include and .dep. As far as I understand .dep should contain modules. Additionally, the order of Makefiles might matter because of FEAUTRES_REQUIRED/PROVIDED, I had some strange errors yesterday.

maribu replied:
You are right. Most code in `cpu` does not do this properly (unlike in `boards` or `drivers`, where this distinction is done consistently). Recently added code (e.g. the ESP32, ESP8266, and RISC-V) does comply with this. Maybe it is a good idea to follow their lead and also start using that distinction.

roberthartung replied:
Looking at other boards, I kept FEATURES_REQUIRED and USEMODULE expansion within Makefile.dep as it seems makes a lot more sense to me to have it in `Makefile.dep`.

maribu commented:
OK, code-wise it almost reached the point I'd ACK (@ZetaR60 and @kYc0o: I'd feel more comfortable if you also could have a brief look at this PR).

The next step would be testing. I added a list of checkmarks above to track the state. I can test on ATmega328p and ATmega2560.

Maybe @Josar can test for ATmega256rfr2?
Maybe @ZetaR60 can test for ATmega1284P?

Regarding the ATmega1281 it will likely go untested, as no one seems to have a waspmote pro any more. (But this will not prevent this PR from merging.)

roberthartung commented:
> OK, code-wise it almost reached the point I'd ACK (@ZetaR60 and @kYc0o: I'd feel more comfortable if you also could have a brief look at this PR).
> 
> The next step would be testing. I added a list of checkmarks above to track the state. I can test on ATmega328p and ATmega2560.
> 
> Maybe @Josar can test for ATmega256rfr2?
> Maybe @ZetaR60 can test for ATmega1284P?
> 
> Regarding the ATmega1281 it will likely go untested, as no one seems to have a waspmote pro any more. (But this will not prevent this PR from merging.)

Looks like I have a bug with the state, but should be easy to fix and we should be good to go for testing then.

maribu commented on boards/common/arduino-atmega/Makefile.dep:
```diff
@@ -3,3 +3,5 @@ USEMODULE += boards_common_atmega
   USEMODULE += saul_gpio
 endif
+
+-include $(RIOTCPU)/$(CPU)/Makefile.dep
```
I think this is better placed in `boards/common/atmega/Makefile.dep`, as not all AVR based boards are Arduinos. (And this and every non-Arduino AVR board should then include `$(RIOTBOARD)/common/atmega/Makefile.dep`).

Also, I would prefer to use `include $(RIOTCPU)/$(CPU)/Makefile.dep` instead of `-include` here, so that developers adding support for new ATmega CPUs will get an error if they did not provide a corresponding `Makefile.dep`.

maribu commented:
I added some remarks how testing can be done in the (a.f.a.i.k.) most straightforward way.

roberthartung commented:
Fun fact: atmega328p does not have *all* banks saturated. Will update shortly. Currently testing on an Arduino Uno!

roberthartung commented:
Some bad things are happening for me:

make BUILD_IN_DOCKER=1 DOCKER="sudo docker" BOARD=arduino-uno all
```
Launching build container using image "riot/riotbuild:latest".
sudo docker run --rm -t -u "$(id -u)" \
    -v '/home/hartung/git/RIOT-ibr:/data/riotbuild/riotbase' \
    -v '/home/hartung/git/RIOT-ibr/build:/data/riotbuild/build' \
    -v '/home/hartung/git/RIOT-ibr/cpu:/data/riotbuild/riotcpu' \
    -v '/home/hartung/git/RIOT-ibr/boards:/data/riotbuild/riotboard' \
    -v '/home/hartung/git/RIOT-ibr/makefiles:/data/riotbuild/riotmake' \
    -v '/home/hartung/git/RIOT-ibr:/data/riotbuild/riotproject' \
    -v '/usr/share/zoneinfo/Europe/Berlin:/etc/localtime:ro' \
    -e 'RIOTBASE=/data/riotbuild/riotbase' \
    -e 'BUILD_DIR=/data/riotbuild/build' \
    -e 'CCACHE_BASEDIR=/data/riotbuild/riotbase' \
    -e 'RIOTCPU=/data/riotbuild/riotcpu' \
    -e 'RIOTBOARD=/data/riotbuild/riotboard' \
    -e 'RIOTMAKE=/data/riotbuild/riotmake' \
    -e 'RIOTPROJECT=/data/riotbuild/riotproject' \
       \
    -e 'BOARD=arduino-uno' \
    -w '/data/riotbuild/riotproject/tests/periph_gpio/' \
    'riot/riotbuild:latest' make all 'BOARD=arduino-uno'
Building application "tests_periph_gpio" for "arduino-uno" with MCU "atmega328p".

"make" -C /data/riotbuild/riotbase/core
"make" -C /data/riotbuild/riotbase/drivers
"make" -C /data/riotbuild/riotbase/drivers/periph_common
"make" -C /data/riotbuild/riotbase/sys
"make" -C /data/riotbuild/riotbase/sys/auto_init
"make" -C /data/riotbuild/riotbase/sys/benchmark
"make" -C /data/riotbuild/riotbase/sys/div
"make" -C /data/riotbuild/riotbase/sys/isrpipe
"make" -C /data/riotbuild/riotbase/sys/shell
"make" -C /data/riotbuild/riotbase/sys/stdio_uart
"make" -C /data/riotbuild/riotbase/sys/tsrb
"make" -C /data/riotbuild/riotbase/sys/xtimer
"make" -C /data/riotbuild/riotboard/arduino-uno
"make" -C /data/riotbuild/riotboard/common/arduino-atmega
"make" -C /data/riotbuild/riotboard/common/atmega
"make" -C /data/riotbuild/riotcpu/atmega328p
"make" -C /data/riotbuild/riotcpu/atmega_common
"make" -C /data/riotbuild/riotcpu/atmega_common/avr_libc_extra
"make" -C /data/riotbuild/riotcpu/atmega_common/periph
/usr/lib/gcc/avr/4.9.2/../../../avr/bin/ld: address 0x800950 of /data/riotbuild/riotproject/tests/periph_gpio/bin/arduino-uno/tests_periph_gpio.elf section `.data' is not within region `data'
/usr/lib/gcc/avr/4.9.2/../../../avr/bin/ld: address 0x800db6 of /data/riotbuild/riotproject/tests/periph_gpio/bin/arduino-uno/tests_periph_gpio.elf section `.bss' is not within region `data'
/usr/lib/gcc/avr/4.9.2/../../../avr/bin/ld: address 0x800db8 of /data/riotbuild/riotproject/tests/periph_gpio/bin/arduino-uno/tests_periph_gpio.elf section `.noinit' is not within region `data'
/usr/lib/gcc/avr/4.9.2/../../../avr/bin/ld: address 0x800950 of /data/riotbuild/riotproject/tests/periph_gpio/bin/arduino-uno/tests_periph_gpio.elf section `.data' is not within region `data'
/usr/lib/gcc/avr/4.9.2/../../../avr/bin/ld: address 0x800db6 of /data/riotbuild/riotproject/tests/periph_gpio/bin/arduino-uno/tests_periph_gpio.elf section `.bss' is not within region `data'
/usr/lib/gcc/avr/4.9.2/../../../avr/bin/ld: address 0x800db8 of /data/riotbuild/riotproject/tests/periph_gpio/bin/arduino-uno/tests_periph_gpio.elf section `.noinit' is not within region `data'
collect2: error: ld returned 1 exit status
/data/riotbuild/riotbase/Makefile.include:460: recipe for target '/data/riotbuild/riotproject/tests/periph_gpio/bin/arduino-uno/tests_periph_gpio.elf' failed
make: *** [/data/riotbuild/riotproject/tests/periph_gpio/bin/arduino-uno/tests_periph_gpio.elf] Error 1
make: *** [/home/hartung/git/RIOT-ibr/makefiles/docker.inc.mk:114: ..in-docker-container] Error 2
```

make BOARD=arduino-uno all
```
Building application "tests_periph_gpio" for "arduino-uno" with MCU "atmega328p".

"make" -C /home/hartung/git/RIOT-ibr/boards/arduino-uno
"make" -C /home/hartung/git/RIOT-ibr/boards/common/arduino-atmega
"make" -C /home/hartung/git/RIOT-ibr/boards/common/atmega
"make" -C /home/hartung/git/RIOT-ibr/core
"make" -C /home/hartung/git/RIOT-ibr/cpu/atmega328p
"make" -C /home/hartung/git/RIOT-ibr/cpu/atmega_common
"make" -C /home/hartung/git/RIOT-ibr/cpu/atmega_common/avr_libc_extra
"make" -C /home/hartung/git/RIOT-ibr/cpu/atmega_common/periph
"make" -C /home/hartung/git/RIOT-ibr/drivers
"make" -C /home/hartung/git/RIOT-ibr/drivers/periph_common
"make" -C /home/hartung/git/RIOT-ibr/sys
"make" -C /home/hartung/git/RIOT-ibr/sys/auto_init
"make" -C /home/hartung/git/RIOT-ibr/sys/benchmark
"make" -C /home/hartung/git/RIOT-ibr/sys/div
"make" -C /home/hartung/git/RIOT-ibr/sys/isrpipe
"make" -C /home/hartung/git/RIOT-ibr/sys/shell
"make" -C /home/hartung/git/RIOT-ibr/sys/stdio_uart
"make" -C /home/hartung/git/RIOT-ibr/sys/tsrb
"make" -C /home/hartung/git/RIOT-ibr/sys/xtimer
/usr/bin/avr-ld: address 0x800956 of /home/hartung/git/RIOT-ibr/tests/periph_gpio/bin/arduino-uno/tests_periph_gpio.elf section `.data' is not within region `data'
/usr/bin/avr-ld: address 0x800dbc of /home/hartung/git/RIOT-ibr/tests/periph_gpio/bin/arduino-uno/tests_periph_gpio.elf section `.bss' is not within region `data'
/usr/bin/avr-ld: address 0x800dbe of /home/hartung/git/RIOT-ibr/tests/periph_gpio/bin/arduino-uno/tests_periph_gpio.elf section `.noinit' is not within region `data'
/usr/bin/avr-ld: address 0x800956 of /home/hartung/git/RIOT-ibr/tests/periph_gpio/bin/arduino-uno/tests_periph_gpio.elf section `.data' is not within region `data'
/usr/bin/avr-ld: address 0x800dbc of /home/hartung/git/RIOT-ibr/tests/periph_gpio/bin/arduino-uno/tests_periph_gpio.elf section `.bss' is not within region `data'
/usr/bin/avr-ld: address 0x800dbe of /home/hartung/git/RIOT-ibr/tests/periph_gpio/bin/arduino-uno/tests_periph_gpio.elf section `.noinit' is not within region `data'
collect2: error: ld returned 1 exit status
make: *** [/home/hartung/git/RIOT-ibr/tests/periph_gpio/../../Makefile.include:460: /home/hartung/git/RIOT-ibr/tests/periph_gpio/bin/arduino-uno/tests_periph_gpio.elf] Error 1
```

maribu commented:
Ah, that is the linker test failing because ROM/RAM requirements exceed the provided falsh/RAM.

(On ARM the message is human readable and also says how much bytes need to be squeezed out of the build to fit the board.) This just means other means of testing are required for ATmega328P based boards

maribu commented:
(Btw: The test was to huge before this PR for the ATmega328p on the docker and my toolchain.)

maribu commented:
The `jiminiy-mega256rfr2`, the `waspmote-pro` and the `mega-xplained` are the currently supported non-Arduino AVR boards and need the `include $(RIOTBOARD)/common/atmega/Makefile.dep` also added to their `Makefile.dep`.

maribu commented:
Perfect. With all my remarks addressed please squash, so that others do not need to dig trough the long commit history.

maribu commented:
Sorry for asking for more changes, but RIOT's internal code should be as efficient as possible to let the application developer have as much resources as possible left. (Also: I just wrote a simple test application that still doesn't fit the ATmega328p. I will really have trouble testing this :-()

roberthartung commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -59,7 +63,101 @@
+/**
+ * @brief
+ */
+static const gpio_t pcint_mapping[][8] = {
```
Might be possible, does it really do much? Isn't it just a notation thing? Memory-wise it should be exactly the same?

maribu replied:
By being able to have a single loop iterating over it instead of two nested loops I'd bet a few instructions are saved. The array itself will have the same size regardless of being 1d or 2d.

roberthartung commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -254,6 +422,66 @@ static inline void irq_handler(uint8_t int_num)
+        /* check if this pin is enabled & has changed */
+        if (enabled_pcints & 0x1) {
+            uint8_t pin_mask = (1 << pin_num);
+            gpio_t pin = pcint_mapping[ bank ][ pin_num ];
```
`*8` for better readability?

maribu commented:
@roberthartung: Do you have an Arduino Uno? With the changes suggest above [this test](https://github.com/maribu/RIOT/commit/16198fd7a701584723d643e6424279f5977c5c8d) for the Arduino Uno can be build and flashed (at least with my toolchain). The test works without module `atmega_pcint` used (for the Arduino pins 2 and 3, as they are the only with their own interrupt vector), but strange crashes occur when the module is added.

maribu commented:
(Btw: Feel free to just cherry-pick the commit adding the test into your PR. It might be useful to have the test around in RIOT.)

roberthartung commented:
> (Btw: Feel free to just cherry-pick the commit adding the test into your PR. It might be useful to have the test around in RIOT.)

How is it a test for Aduino Uno if your makefile says `BOARD_INSUFFICIENT_MEMORY := arduino-duemilanove arduino-uno`?

roberthartung commented:
> @roberthartung: Do you have an Arduino Uno? With the changes suggest above [this test](https://github.com/maribu/RIOT/commit/16198fd7a701584723d643e6424279f5977c5c8d) for the Arduino Uno can be build and flashed (at least with my toolchain). The test works without module `atmega_pcint` used (for the Arduino pins 2 and 3, as they are the only with their own interrupt vector), but strange crashes occur when the module is added.

Output for melooks fine so far.

roberthartung commented:
Strange.. for me the flank_name does not match.

maribu commented:
> How is it a test for Aduino Uno if your makefile says BOARD_INSUFFICIENT_MEMORY := arduino-duemilanove arduino-uno?

Sorry, an artifact of copy pasting the Makefile from a different test :-(

maribu commented:
> Strange.. for me the flank_name does not match.

Keep in mind that the test could be buggy as well...

> Output for me looks fine so far.

Maybe the issue is the reduces stack size and my toolchain creates less efficient code then yours? Did you use the docker toolchain, if so I should give that a try, too.

maribu commented:
> Strange.. for me the flank_name does not match.

By that you mean that it is surprising that you have a single event on the console for a pin with both flanks enabled?

roberthartung commented:
> > Strange.. for me the flank_name does not match.
> 
> By that you mean that it is surprising that you have a single event on the console for a pin with both flanks enabled?

the configuring is on "both" but printing the flank name gives me "falling", something is wrong here. Will continue on monday!

maribu commented:
> Will continue on monday!

:+1:

Have a nice weekend!

roberthartung commented:
Very strange ... the pcint handler in `gpio.c` seems to match the correct bank + pin offset, but it breaks on a higher level. Will keep debugging:

```
2019-03-11 10:05:37,482 - INFO # pcint for 2/7
2019-03-11 10:05:37,536 - INFO # INT: 6 Pin = 8; IRQ type = PCINT0; flank(s) = falling
2019-03-11 10:05:37,789 - INFO # pcint for 2/7
2019-03-11 10:05:37,843 - INFO # INT: 6 Pin = 8; IRQ type = PCINT0; flank(s) = falling
2019-03-11 10:05:38,572 - INFO # pcint for 2/6
2019-03-11 10:05:38,626 - INFO # INT: 6 Pin = 8; IRQ type = PCINT0; flank(s) = falling
2019-03-11 10:05:39,038 - INFO # pcint for 2/5
2019-03-11 10:05:39,096 - INFO # INT: 6 Pin = 8; IRQ type = PCINT0; flank(s) = falling
2019-03-11 10:05:39,767 - INFO # pcint for 2/4
2019-03-11 10:05:39,825 - INFO # INT: 6 Pin = 8; IRQ type = PCINT0; flank(s) = falling
2019-03-11 10:05:40,042 - INFO # pcint for 2/4
2019-03-11 10:05:40,100 - INFO # INT: 6 Pin = 8; IRQ type = PCINT0; flank(s) = falling
2019-03-11 10:05:40,505 - INFO # INT: 1 Pin = 3; IRQ type = INT1; flank(s) = rising
2019-03-11 10:05:40,558 - INFO # INT: 1 Pin = 3; IRQ type = INT1; flank(s) = rising
2019-03-11 10:05:40,783 - INFO # INT: 1 Pin = 3; IRQ type = INT1; flank(s) = rising
2019-03-11 10:05:40,836 - INFO # INT: 1 Pin = 3; IRQ type = INT1; flank(s) = rising
2019-03-11 10:05:41,258 - INFO # INT: 0 Pin = 2; IRQ type = INT0; flank(s) = falling
2019-03-11 10:05:41,316 - INFO # INT: 0 Pin = 2; IRQ type = INT0; flank(s) = falling
```

Found the bug! :-)

`conf->cb(pcint_config->arg);` -> `conf->cb(conf->arg);`

I wonder why this wasn't a compiler error in the first place?!

roberthartung commented:
@maribu Feel free and go on with testing now :)

maribu commented:
Works as expected on the Arduino UNO :-)

roberthartung commented:
It looks like the `ATmega256RFR2` has three banks, but only 9 actual usable PCints, do you agree @maribu ?

maribu commented:
OK, I double checked the configuration of the ATmega256RFR2 and the ATmega1281 against the data sheet. As I doubt we will receive testing for those boards, I would say this is as good as we can do for those.

roberthartung commented:
@TorbenPetersen Can you check if it works for INGA (atmega1284p) and give feedback here?

maribu commented:
@roberthartung: Interrupts on the Arudino Mega2560 don't work for me. I [updated the test](https://github.com/maribu/RIOT/commit/85cda485715843924b0ee00df2aa0ad270175dc1) to work on both Uno and Mega. It seems to me that there is a general issue with interrupts on the Mega2560 that is unrelated to this PR.

roberthartung commented:
cherry-picked your test and squashed the commits.

TorbenPetersen commented:
> @TorbenPetersen Can you check if it works for INGA (atmega1284p) and give feedback here?

I tested your code on the INGa Board. The pcint seems to work.

maribu commented:
OK. I'd say IRQ for the Mega2560 needs to be fixed regardless of this PR. As this PR only adds to the compiled binary if the corresponding module is manually added, there is no risk of breaking stuff. (Worst thing that could happen is that the new feature does not work for the `jiminy-mega256rfr2` and the `waspmote-pro`.)

Two things I'd like to see:
1. Splitting the commit into two: One for the feature, one for the test
2. Adding some documentation. (I'd say put it once in `boards/common/atmega/doc.txt`, instead of copy pasting it into every board. But I'd put a link to that page in every ATmega board so that people actually end up reading the doc.)

maribu commented on boards/common/arduino-atmega/Makefile.dep:
```diff
@@ -3,3 +3,5 @@ USEMODULE += boards_common_atmega
   USEMODULE += saul_gpio
 endif
+
+include $(RIOTBOARD)/common/atmega/Makefile.dep
```
Missing `\n` here. Maybe this is why Murdock fails to parse Makefiles?

roberthartung replied:
Last time I added `\n` gave me some additional new line errors-error?

maribu commented on boards/common/atmega/Makefile.dep:
```diff
@@ -0,0 +1 @@
+include $(RIOTCPU)/$(CPU)/Makefile.dep
```
Missing `\n` here.

maribu commented on boards/common/atmega/Makefile.include:
```diff
@@ -1 +1,4 @@
 INCLUDES += -I$(RIOTBOARD)/common/atmega/include
+
+# require Makefile.dep for correct dependency management of atmega PCINTs
+include $(RIOTCPU)/$(CPU)/Makefile.dep
```
Missing `\n` here.

maribu replied:
This file is already included via `boards/common/atmega/Makefile.dep`

maribu replied:
I thin this is the issue with Murdock. I believe the CPU variable is not yet available here. Also: `boards/common/atmega/Makefile.dep` already contains this exact include, so no need to include it twice.

roberthartung commented:
> OK. I'd say IRQ for the Mega2560 needs to be fixed regardless of this PR. As this PR only adds to the compiled binary if the corresponding module is manually added, there is no risk of breaking stuff. (Worst thing that could happen is that the new feature does not work for the `jiminy-mega256rfr2` and the `waspmote-pro`.)
> 
> Two things I'd like to see:
> 
> 1. Splitting the commit into two: One for the feature, one for the test
> 2. Adding some documentation. (I'd say put it once in `boards/common/atmega/doc.txt`, instead of copy pasting it into every board. But I'd put a link to that page in every ATmega board so that people actually end up reading the doc.)

What kind of link should I put in each doc?

maribu commented on cpu/atmega1281/Makefile.dep:
```diff
@@ -0,0 +1 @@
+-include $(RIOTCPU)/atmega_common/Makefile.dep
```
Missing `\n` here.

maribu replied:
This could be `include` instead of `-include`

maribu commented on cpu/atmega1284p/Makefile.dep:
```diff
@@ -0,0 +1,8 @@
+
+# FEATURE_REQUIRED for atmega_pcint3 centralized in atmega_common for future use cases!
+
+-include $(RIOTCPU)/atmega_common/Makefile.dep
```
Missing `\n` here.

maribu commented on cpu/atmega256rfr2/Makefile.dep:
```diff
@@ -0,0 +1 @@
+-include $(RIOTCPU)/atmega_common/Makefile.dep
```
Missing `\n` here.

maribu commented on cpu/atmega328p/Makefile.dep:
```diff
@@ -0,0 +1 @@
+-include $(RIOTCPU)/atmega_common/Makefile.dep
```
Missing `\n` here.

maribu commented on cpu/atmega_common/Makefile.dep:
```diff
@@ -0,0 +1,24 @@
+# feature for atmega_pcint3 module
+ifneq (,$(filter atmega_pcint3,$(USEMODULE)))
+    FEATURES_REQUIRED += atmega_pcint3
+endif
```
Missing `\n` here.

maribu commented:
E.g. add

``` Doxygen
## Pin Change Interrupts

More pins can be used for hardware interrupts using the Pin Change
Interrupt feature. See @ref boards_common_atmega for details.
```

maribu commented:
Unrelated to the Murdock issue: Now some lines have a blank line at the end (that is two `\n` after the last non white-space character). This is something the CI doesn't like as well. (Honestly: In my opinion the text editor should handle that one `\n` is added to the last line. E.g. `vim` handles this correctly.)

roberthartung commented:
> Unrelated to the Murdock issue: Now some lines have a blank line at the end (that is two `\n` after the last non white-space character). This is something the CI doesn't like as well. (Honestly: In my opinion the text editor should handle that one `\n` is added to the last line. E.g. `vim` handles this correctly.)

Well .. I used vim on exactly the files you told me to change.. strange ;)

maribu commented:
Try `:set numbers` to display line numbers. The should be no line after the last line with content. `vim` will terminate the last line transparently and correctly with `\n`.

maribu commented:
(so basically opening and saving the files with vim without changing anything would have fixed the issue.)

roberthartung commented:
> (so basically opening and saving the files with vim without changing anything would have fixed the issue.)

Removed the white spaces again with the last fixup. Looks exactly the same for me as before?!

maribu commented:
No, now it is correct. (Have a look [at this commit](https://github.com/RIOT-OS/RIOT/pull/11122/commits/b2ca44486a1c0a644b78f1b4e6144fb6b85d6bf5) at the end of the file `boards/common/arduino-atmega/Makefile.dep`: Github adds a red icon to indicate the missing `\n`.)

roberthartung commented:
squashed and split up. :)

maribu commented on cpu/atmega_common/Makefile.dep:
```diff
@@ -0,0 +1,24 @@
+# expand atmega_pcint module
+ifneq (,$(filter atmega_pcint,$(USEMODULE)))
+    USEMODULE += atmega_pcint0 atmega_pcint1 atmega_pcint2
```
With only `atmega_pcint0` and `atmega_pcint1` now being supported on all AVRs, this needs to be adapted. Also: The indent here needs to be two spaces, instead of four. (The Makefiles have a different code style than the C files.)

How about this:

```C
ifneq (,$(filter atmega_pcint,$(USEMODULE)))
  USEMODULE += atmega_pcint0 atmega_pcint1
  ifneq (,$(filter atmega_pcint2,$(FEATURES_PROVIDED)))
    USEMODULE += atmega_pcint2
  endif
  ifneq (,$(filter atmega_pcint3,$(FEATURES_PROVIDED)))
    USEMODULE += atmega_pcint2
  endif
endif
```

But that will only work if the features are already in place when this is evaluated. But as the common Makefile is included from the CPU specific ones, this should be true.

maribu commented:
@roberthartung: While trying to figure out the issue that causes Murdock to give up, I added some commits. (I also messed up to pass the right commit ID to `--fixup=`, the last commit addresses an issue in the first commit of this PR.) Please note that 1 level of indent is 2 spaces in a Makefile, not 4. (But it is indeed 4 spaces in C source and header files.)

@kaspar030: I have no idea what is going wrong here. Murdock fails to parse the Makefiles of all exmaples and tests that have not blacklisted one or more ATmega based boards. This PR adds/changes Makefiles for ATmega based boards, so likely there is some issue. But I cannot find it. Could you help?

kaspar030 commented on boards/common/atmega/Makefile.dep:
```diff
@@ -0,0 +1,2 @@
+# require Makefile.dep for correct dependency management of atmega PCINTs
+include $(RIOTCPU)/$(CPU)/Makefile.dep
```
This line is the problem. ```${BOARD}``` and ```${CPU}``` are not known when executing the ```info-boards-supported``` target, which is used to find out which boards an application supports.

kaspar030 replied:
See this:

```
[kaspar@ng riot.tmp ((ff3e541c09...))]$ make -C tests/xtimer_usleep info-boards-supported
make: Entering directory '/home/kaspar/src/riot.tmp/tests/xtimer_usleep'
/home/kaspar/src/riot.tmp/boards/common/atmega/Makefile.dep:2: /home/kaspar/src/riot.tmp/cpu//Makefile.dep: No such file or directory
make: *** No rule to make target '/home/kaspar/src/riot.tmp/cpu//Makefile.dep'.  Stop.
make: Leaving directory '/home/kaspar/src/riot.tmp/tests/xtimer_usleep'
[kaspar@ng riot.tmp ((ff3e541c09...))]$
```

maribu commented:
OK, I was able to force push. I used that to add again the commit I messed up, but this time with the correct commit id passed to `--fixup=<ID>`. The last commit pushed should hopefully address the issue @kaspar030 thankfully explained. Now the Murdock build (except for the missing squashing) passes :-)

@roberthartung: Please check the commits I added on top of yours. If you agree, please squash.

maribu commented:
@kaspar030, @kYc0o, @ZetaR60: This feature has been quite controversial in an earlier PR. Thus, I'd like to have a second ACK before merging. In the last PR there were two major points of critique:

1. Conflicts with the implementation of context switches back then, as the code at that point in time used PCINT0 interrupt for context switches. In the meantime (and for unrelated reasons) the context switch code was changed to no longer use PCINT0 for that. So this point no longer applies
2. Increase in RAM and ROM

With this PR using a pseudo module, RAM and ROM requirements are not increased unless someone explicitly unlocks this feature. I also believe that there is little room for improvement regarding RAM and ROM size - at least not without using arcane magic like carefully crafted inline assembly.

Btw: It is very nice that this PR allows to use the hardware independent API `gpio_init_int()` and completely shields the application programmer from the weird shared interrupt vectors using here.

roberthartung commented:
@maribu squashed and rebased to latest master. Is there a documentation about the Makefile order somewhere? If not we should create one because that was quite hard to understand in the first place.

maribu commented:
Good idea. Maybe @kaspar030 could also add "try running make info-boards-supported to see whats wrong" to the error message in Murdock, so that people see how to reproduce the issue locally.

kaspar030 commented:
> Maybe @kaspar030 could also add "try running make info-boards-supported to see whats wrong" to the error message in Murdock, so that people see how to reproduce the issue locally.

Yeah. Please see #11160 (and it's current CI output). I guess that output would've saved some time?

roberthartung commented:
@maribu So .. how's the plan? Is there anything missing?

maribu commented:
Currently @kaspar030 request for changes is blocking this PR from being merged. We should wait for him to review the changes and confirm that his comments are addressed. (As his comments targeted the bug that prevented Murdock from building successfully and Murdock now did build successfully, I would say those comments have been addressed.)

Also: As this feature was extremely controversial, I marked this PR to need at least two ACKs. I think the reason why this feature was controversial have been addressed (as I [explained in detail a few comments above](https://github.com/RIOT-OS/RIOT/pull/11122#issuecomment-471729008)). But it makes still sense to me that someone else agrees on that before merging.

maribu commented:
@kYc0o, @ZetaR60, @kaspar030: Ping?

maribu commented:
OK, regarding the Arduino-Mega2560: I just fucked up the test. Everything but PCINT9 and PCINT10 is working there. (PCINT9 and PCINT10 are btw the only two PCINT pins of that PCINT bank available there, so maybe the whole bank is not working?)

Here is a fixed version of the test I used. @roberthartung: Do you have an Arduino Mega2560 at hand so that you can reproduce? Can you fix the test in this PR accordingly? (Feel free to squash right away.)

roberthartung commented:
@maribu Are you sure about the PCINTs? The data sheet says it has full three banks.

maribu commented:
@roberthartung: Sorry for being inprecise. The ATmega2560 has three full banks. But not all pins of bank two on the Arduino Mega2560 are connected to any of the pin headers. Also the RX0/TX0 pins are not tested as those are used for the stdio uart. (In case of the Mega only one of them was actually a pcint pin, if I remember correctly.) For those reasons I could only test two of the 8 pins on bank two on the Arduino Mega.

maribu commented:
@roberthartung: I'm extremely sorry for delaying this PR. I honestly thought that interrupt handling on the Mega2560 was broken anyway and was heading to fix this in Master, but then noticed that only my test was broken.

Could you reproduce the issue on the Arduino Mega2560? Maybe my board is just broken.

maribu commented:
@roberthartung How about just disabling the pcint on bank two specifically for the ATmega2560? The only ATmega2560 based board in RIOT only has two of the 8 pins of bank two available anyway, so it is not extremely useful anyway. With that, this would no longer block this PR.

@kaspar030: Could you dismiss your request for changes? Murdock is now running, so your remarks have been addressed.

roberthartung commented:
@maribu Looking into this right now. Why did you say that only some of the PCINTs are usable on Arduino-mega2560? Easiest way would be to remove mega2560 completely for now?

roberthartung commented:
Re-based and added pcint header file for arduino mega2560. I'm wondering if we should move the header files to each board @maribu, as the CPU always supports everything, but the difference is in boards? Or should we stick to the "overriding" method?

maribu commented:
Of the second PCINT Port (PCINT8 ... PCINT15) the pins PE0 (PCINT8, RXD0), PJ0 (PCINT9, RXD3) and PJ1 (PCINT10/TXD3) are available on the Arduino Mega. PJ2 - PJ6 (PCINT11 - PCINT15) are not mapped.

I did not test PE0, as UART0 is used for stdio. I did expect PJ0 and PJ1 to work, as only UART 0 and UART 1 are configured in `boards/common/atmega/include/periph_conf_atmega_common.h`, and UART 2 and UART 3 remain unused. Maybe there is indeed a conflict with UART 3 due to something I did overlook. I'll try to investigate.

I would personally keep the pcint configuration in the CPU and keep it fully populated in order to keep the maintenance burden low. (E.g. `gpio_init()` does also not test for GPIO pins that are available on the MCU, but not on the board.)

May point was that I could not confirm that the pcint implementation on port 2 of the ATmega2560 is working, which blocked this PR. Disabling that port for now (with the possibility of adding it later on) would allow me to set testing on green.

But a conflict with UART seems to be indeed the likely cause for the issue here. I hope that I can have a look at that again soon.

maribu commented:
@roberthartung: The fix of the unit test in https://github.com/maribu/RIOT/commits/pcint have not been integrated yet. Care to do so?

maribu commented:
I updated the test to also add pins not exposed by the Arduino Mega2560. When removing the `atmega_pcint.h` in the Arduino Mega board I can successfully test the pin changed interrupts with a jumper cable directly at the chip.

maribu commented:
I'd say update the test and then merge. (Sorry for the long stall.)

maribu commented:
This PR does not change the behavior nor the ROM/RAM requirements nor the binaries of any existing user, unless the pseudo module is explicitly pulled in. Therefore, this is not an `>1 ACK` PR.

Also, the other reviewers had enough time to complain. If they were really upset with the approach of using a pseudo module, they would have said so by now.

roberthartung commented:
@maribu I cherry picked your commit and updated the branch. I guess we're finally good to go? :)

maribu commented:
> I guess we're finally good to go? :)

I'll do one more round of testing tomorrow to make double sure everything is ready, but I'd say yes :-)

maribu commented on boards/arduino-mega2560/include/atmega_pcint.h:
```diff
@@ -0,0 +1,16 @@
+}
+#endif
+
+#endif /* ATMEGA_PCINT_H */
```
This file now conflicts with the test: The test sets up also PCINT interrupts for pins not connected to the Arduino Mega2560 header. (If I recall correctly it was my suggestion to override the pcint mapping for this board. But now I think this was a bad idea. E.g. other RIOT code does not prevent the user from using pins available at the MCU but not connected to pin header. Maybe it would be better to remove this file and stick with the default mapping for the ATmega2560 MCU?)

roberthartung replied:
What is the actual conflict here? If the pins are not correct for that board, I think it's totally fine to override the pcint definition. The test should then not setup pcints that are not available. Or better: the test should expect the gpio_init_int() call to fail, if the pin is invalid on the board.

maribu replied:
> Or better: the test should expect the gpio_init_int() call to fail, if the pin is invalid on the board. 

+1

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -254,6 +428,77 @@ static inline void irq_handler(uint8_t int_num)
+                }
+                else {
+                    pcint_state[ bank ] &= ~(pin_mask);
+                }
```
```C
                if (pin_value) {
                    pcint_state[ bank ] |= (pin_mask);
                }
                else {
                    pcint_state[ bank ] &= ~(pin_mask);
                }
```
This could be simplified to `pcint_state[bank] ^= pin_mask;`, which will just toggle the bit. That would be correct for both cases

roberthartung replied:
The idea is *not* to toggle the bit, but to set the bit correctly to 1 or 0. Using your syntax does not work here!

maribu replied:
This is only entered if the current value of the pin is different from the value read from the pin differs from the value stored in `pcint_state`.

If `pin_value` is not 0 (then-case of the inner `if`) and also `pin_value` differs from what it is stored (then-case of outer `if`), then the pin is set. If both `then`-branches are taken, the value in `pcint_state` for that bit has to be zero and is not set. This is equivalent to toggling.

if `pin_value` is zero (else-case of inner `if`) and differs from what is stored in `pcint_state` (then-case of outer if), than the bit is clear. But to take the then-branch of the outer `if` and the else-branch of the inner `if`, the corresponding bit had to be `1`. So this is again equivalent to toggling the bit.

In both cases the bit is just toggled, so the code could be written like that to be more efficient. (If you don't believe me: Change it to toggling and run the test ;-))

maribu replied:
I fixed to typos above that made understanding what I meant impossible. It should now make sense.

roberthartung replied:
In the latter case for the `!=` I agree, then we can toggle. For the initial state you can't!

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -254,6 +428,77 @@ static inline void irq_handler(uint8_t int_num)
+        /* check if this pin is enabled & has changed */
+        if (enabled_pcints & 0x1) {
+            /* re-construct mask from pin number*/
+            uint8_t pin_mask = (1 << pin_num);
```
This does not work, as `pin_num` is the idx in the PCINT mapping, not the pin number in the GPIO port. (Those happen to be the same for all but the ATmega2560.)

Here is my suggestion to fix this: https://github.com/maribu/RIOT/commit/8f28ebd73272a4dac4b112b6fa2a3ad8c348ddd3

roberthartung replied:
done.

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -254,6 +428,77 @@ static inline void irq_handler(uint8_t int_num)
 
+#ifdef PCINT_NUM_BANKS
+/* inline function that is used by the PCINT ISR */
+static inline void pcint_handler(uint8_t bank, volatile uint8_t *mask_reg)
```
instead of `volatile uint8_t *mask_reg`, why not directly pass the value as `uint8_t enabled_pcints` in this function?

roberthartung replied:
True. Good point.

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -254,6 +428,77 @@ static inline void irq_handler(uint8_t int_num)
+#if defined(PCINT0_IDX)
+ISR(PCINT0_vect, ISR_BLOCK)
+{
+    pcint_handler(PCINT0_IDX, &PCMSK0);
```
With my suggestion above, this would become

``` C
cint_handler(PCINT0_IDX, PCMSK0);
```

instead of

``` C
cint_handler(PCINT0_IDX, &PCMSK0);
```

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -254,6 +428,77 @@ static inline void irq_handler(uint8_t int_num)
+#if defined(PCINT1_IDX)
+ISR(PCINT1_vect, ISR_BLOCK)
+{
+    pcint_handler(PCINT1_IDX, &PCMSK1);
```
same as above

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -254,6 +428,77 @@ static inline void irq_handler(uint8_t int_num)
+#if defined(PCINT2_IDX)
+ISR(PCINT2_vect, ISR_BLOCK)
+{
+    pcint_handler(PCINT2_IDX, &PCMSK2);
```
same as above

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -254,6 +428,77 @@ static inline void irq_handler(uint8_t int_num)
+#if defined(PCINT3_IDX)
+ISR(PCINT3_vect, ISR_BLOCK)
+{
+    pcint_handler(PCINT3_IDX, &PCMSK3);
```
same as above

maribu commented:
OK, apart from the the test (which should expect `gpio_init_int()` to fail on pins not supported by a board) I'd say ACK. How about splitting the test off in a separate PR, squash this PR and merge?

To me it also is questionable if I am actually allowed to ACK the code of the test, as I authored most of it. Maybe I can open a separate PR for that and you (@roberthartung) could review the test?

maribu commented:
Oh, I forgot that ATmega32U4 has been merged in the meantime. Can you please rebase and add the pcint mapping for that one? I will test on the Arduino Leonardo.

maribu commented on cpu/atmega_common/periph/gpio.c:
```diff
@@ -195,8 +301,80 @@ int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
+        switch (bank) {
+#ifdef PCINT0_IDX
+            case PCINT0_IDX:
+                PCMSK0 |= (1 << pin_num);
```
Here again a bug: The pin number and the index in the PCINT bank do not match for the ATmega2560. I fixed the issue locally in https://github.com/RIOT-OS/RIOT/pull/11923/commits/8f669ae86f9e87d9a98a7094020a810c014e5433. With that everything works fine :-)

roberthartung replied:
Ok, cherry picked!

maribu commented on cpu/atmega_common/Makefile.features:
```diff
@@ -1,2 +1,4 @@
 FEATURES_PROVIDED += periph_eeprom
 FEATURES_PROVIDED += periph_pm
+
+FEATURES_PROVIDED += atmega_pcint0 atmega_pcint1
```
The ATmega32U4 only as PCINT0. Also, this forces every newly added implementation of ATmega CPUs to also add PCINT right away. Maybe it is better to let each individual ATmega CPU define the list of features provided?

maribu replied:
That would by the way also allow to split up the code changes into different commits, e.g. first adding the implementation in one commit, that add the configuration for each CPU as separate commit, than a final commit adding the Arduino-Mega2560 board-specific configuration.

I think this would help people porting new ATmega CPUs to RIOT, as they could see in a commit adding the PCINT config for a CPU what exactly has to be done.

roberthartung replied:
FEATURES_PROVIDED does not add the module, right? It should shows that the features is available - correct me if I'm wrong here!

maribu replied:
That's correct

roberthartung commented:
> OK, apart from the the test (which should expect `gpio_init_int()` to fail on pins not supported by a board) I'd say ACK. How about splitting the test off in a separate PR, squash this PR and merge?
> 
> To me it also is questionable if I am actually allowed to ACK the code of the test, as I authored most of it. Maybe I can open a separate PR for that and you (@roberthartung) could review the test?

Sounds good for me, just removed the test from my code.

roberthartung commented:
rebased on latest master, one commit for basic implementation, one commit for each CPU. I think having more commits makes it clearer! Additionally I fixed a bug for atmega1284p (missing .dep file) and I added the atmega256rfr2 CPU.

maribu commented:
@roberthartung: Thanks for your patience!
